src/esp/geo/VoxelGrid.h

```diff
@@ -0,0 +1,392 @@
+#include <Magnum/Image.h>
+#include <Magnum/Math/Color.h>
+#include <Magnum/Math/CubicHermite.h>
+#include <Magnum/Math/Range.h>
+#include <Magnum/PixelFormat.h>
+
+#include "esp/assets/MeshData.h"
+#include "esp/assets/MeshMetaData.h"
+#include "esp/core/esp.h"
+#include "esp/geo/geo.h"
+#include "esp/gfx/Drawable.h"
+#include "esp/gfx/magnum.h"
+
+#ifdef ESP_BUILD_WITH_VHACD
+#include "VHACD.h"
+#endif
+
+namespace Mn = Magnum;
+namespace Cr = Corrade;
+
+namespace esp {
+namespace geo {
+
+enum class VoxelGridType { Bool, Int, Float, Vector3 };
+
+class VoxelGrid {
+  struct GridEntry {
+    VoxelGridType type;
+    Cr::Containers::Array<char> data;
+    Cr::Containers::StridedArrayView3D<void> view;
+  };
+
+ public:
+#ifdef ESP_BUILD_WITH_VHACD
+  /**
+   * @brief Generates a Boundary voxel grid using VHACD's voxelization
+   * framework.
+   * @param MeshData The mesh that will be voxelized
+   * @param renderAssetHandle The handle for the render asset.
+   * @param resolution The approximate number of voxels in the voxel grid.
+   */
+  VoxelGrid(const std::unique_ptr<assets::MeshData>& meshData,
+            const std::string& renderAssetHandle,
+            int resolution);
+#endif
+
+  /**
+   * @brief Generates an empty voxel grid given some voxel size and voxel
+   * dimensions..
+   * @param voxelSize The size of a single voxel
+   * @param VoxelGridDimensions The dimensions of the voxel grid.
+   */
+  VoxelGrid(const Mn::Vector3& voxelSize,
+            const Mn::Vector3i& voxelGridDimensions);
+
+  /**
+   * @brief Gets the enumerated type of a particular voxel grid type.
+   * @return The enumerated type.
+   */
+  template <typename T>
+  VoxelGridType voxelGridTypeFor();  // no definition here
+
+  /**
+   * @brief Generates a new empty voxel grid of a specified type.
+   * @param gridName The key under which the grid will be registered and
+   * accessed.
+   */
+  template <typename T>
+  void addGrid(const std::string& gridName) {
+    VoxelGridType type = voxelGridTypeFor<T>();
+
+    std::size_t dims[3]{static_cast<std::size_t>(m_voxelGridDimensions[0]),
+                        static_cast<std::size_t>(m_voxelGridDimensions[1]),
+                        static_cast<std::size_t>(m_voxelGridDimensions[2])};
+
+    Cr::Containers::StridedDimensions<3, std::ptrdiff_t> strides{
+        static_cast<std::ptrdiff_t>(m_voxelGridDimensions[2] *
+                                    m_voxelGridDimensions[1] * sizeof(T)),
+        static_cast<std::ptrdiff_t>(m_voxelGridDimensions[2] * sizeof(T)),
+        static_cast<std::ptrdiff_t>(sizeof(T))};
+
+    if (grids_.find(gridName) != grids_.end()) {
+      // grid exists, simply overwrite
+      Mn::Debug() << gridName << "exists, overwriting.";
+
+      grids_[gridName].data = Corrade::Containers::Array<char>(
+          Corrade::Containers::ValueInit, gridSize() * sizeof(T));
+
+      Cr::Containers::StridedArrayView<3, void> view{grids_[gridName].data,
```

Comment:
FYI you can find&replace all `Cr::Containers::StridedArrayView<3, T>` with `Cr::Containers::StridedArrayView3D<T>`.